core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -720,15 +728,24 @@ private static SortOrder freshSortOrder(int orderId, Schema schema, SortOrder so
     return builder.build();
   }
 
-  private Map<String, SnapshotRef> validateRefs(Map<String, SnapshotRef> inputRefs) {
-    for (Map.Entry<String, SnapshotRef> refEntry : inputRefs.entrySet()) {
-      String refName = refEntry.getKey();
-      SnapshotRef ref = refEntry.getValue();
-      if (!(ref.snapshotId() == -1 && refName.equals(SnapshotRef.MAIN_BRANCH))) {
-        Preconditions.checkArgument(snapshotsById.containsKey(ref.snapshotId()),
-            "Snapshot reference %s does not exist in the existing snapshots list", ref);
-      }
+  private static Map<String, SnapshotRef> validateRefs(Long currentSnapshotId,
+                                                       Map<String, SnapshotRef> inputRefs,
+                                                       Map<Long, Snapshot> snapshotsById) {
+    for (SnapshotRef ref : inputRefs.values()) {
+      Preconditions.checkArgument(snapshotsById.containsKey(ref.snapshotId()),
+          "Snapshot for reference %s does not exist in the existing snapshots list", ref);
     }
+
+    SnapshotRef main = inputRefs.get(SnapshotRef.MAIN_BRANCH);
+    if (currentSnapshotId != -1) {
+      Preconditions.checkArgument(main == null || currentSnapshotId == main.snapshotId(),
+          "Current snapshot ID does not match main branch (%s != %s)",
```

Comment:
nit: I wonder if the message would be slightly clearer like this: `Current snapshot ID '%s' does not match the snapshot ID 
 '%s' of the main branch`

Reply:
I strongly prefer keeping the error message itself readable and adding context after. While it is more clear which ID is which in your version, I think it makes the actual problem harder to read because there are longs embedded in the text.